### PR TITLE
[Merged by Bors] - feat: HahnSeries smul order inequality

### PR DESCRIPTION
**Pull request by ScottCarnahan**
Given a Hahn series `x` and a scalar `r` such that `r • x ≠ 0`, the order of `r • x` is not strictly less than the order of `x`.  If the exponent poset is linearly ordered, the order of `x` is less than or equal to the order of `r • x`.  (Note that the order of a Hahn series is not uniquely defined when the exponent poset is not linearly ordered.) This is a complement to the addition result `HahnSeries.min_order_le_order_add`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/HahnSeries.lean [hunk 1]
```diff
@@ -512,6 +512,14 @@ instance : DistribMulAction R (HahnSeries Γ V) where
     ext
     simp [mul_smul]
 
+theorem smul_order_leq {Γ} [Zero Γ] [LinearOrder Γ] (r : R) (x : HahnSeries Γ V) (h : r • x ≠ 0) :
+    x.order ≤ (r • x).order := by
+  by_cases hx : x = 0
```
Comment (urkud) [inline Mathlib/RingTheory/HahnSeries.lean:517]: You can use `right_ne_zero_of_smul` here.

Comment (ScottCarnahan) [reply Mathlib/RingTheory/HahnSeries.lean:517]: I tried that and found that the statement of `right_ne_zero_of_smul` unnecessarily requires `[Zero R]`.  Should I change it?

Comment (ScottCarnahan) [reply Mathlib/RingTheory/HahnSeries.lean:517]: I have removed the hypothesis and moved it to `GroupTheory.GroupAction.Defs`, but I didn't know what to do with the aligns.

File: Mathlib/RingTheory/HahnSeries.lean [hunk 2]
```diff
@@ -512,6 +512,14 @@ instance : DistribMulAction R (HahnSeries Γ V) where
     ext
     simp [mul_smul]
 
+theorem smul_order_leq {Γ} [Zero Γ] [LinearOrder Γ] (r : R) (x : HahnSeries Γ V) (h : r • x ≠ 0) :
```
[the remark this answers is not in the document]
Comment (ScottCarnahan) [reply Mathlib/RingTheory/HahnSeries.lean]: I added a version that doesn't have `LinearOrder`.

File: Mathlib/Algebra/SMulWithZero.lean [hunk 3]
```diff
@@ -75,12 +75,10 @@ variable {R} {a : R} {b : M}
 
 lemma smul_eq_zero_of_left (h : a = 0) (b : M) : a • b = 0 := h.symm ▸ zero_smul _ b
 #align smul_eq_zero_of_left smul_eq_zero_of_left
-lemma smul_eq_zero_of_right (a : R) (h : b = 0) : a • b = 0 := h.symm ▸ smul_zero a
-#align smul_eq_zero_of_right smul_eq_zero_of_right
+--#align smul_eq_zero_of_right smul_eq_zero_of_right
```
Comment (urkud) [inline Mathlib/Algebra/SMulWithZero.lean:78]: No need to keep them here.
```suggestion
```

File: Mathlib/Algebra/SMulWithZero.lean [hunk 4]
```diff
@@ -75,12 +75,10 @@ variable {R} {a : R} {b : M}
...
 lemma left_ne_zero_of_smul : a • b ≠ 0 → a ≠ 0 := mt fun h ↦ smul_eq_zero_of_left h b
 #align left_ne_zero_of_smul left_ne_zero_of_smul
-lemma right_ne_zero_of_smul : a • b ≠ 0 → b ≠ 0 := mt <| smul_eq_zero_of_right a
-#align right_ne_zero_of_smul right_ne_zero_of_smul
+--#align right_ne_zero_of_smul right_ne_zero_of_smul
```
Comment (urkud) [inline Mathlib/Algebra/SMulWithZero.lean:81]:
```suggestion
```

File: Mathlib/RingTheory/HahnSeries.lean [hunk 5]
```diff
@@ -512,6 +512,16 @@ instance : DistribMulAction R (HahnSeries Γ V) where
     ext
     simp [mul_smul]
 
+theorem smul_order_not_lt [Zero Γ] (r : R) (x : HahnSeries Γ V) (h : r • x ≠ 0) :
```
Comment (urkud) [inline Mathlib/RingTheory/HahnSeries.lean:515]:
```suggestion
theorem order_smul_not_lt [Zero Γ] (r : R) (x : HahnSeries Γ V) (h : r • x ≠ 0) :
```

File: Mathlib/RingTheory/HahnSeries.lean [hunk 6]
```diff
@@ -512,6 +512,16 @@ instance : DistribMulAction R (HahnSeries Γ V) where
...
+    ¬ (r • x).order < x.order := by
+  by_cases hx : x = 0
+  · exact ((right_ne_zero_of_smul h) hx).elim
```
Comment (urkud) [inline Mathlib/RingTheory/HahnSeries.lean:518]:
```suggestion
  have hx : x ≠ 0 := right_ne_zero_of_smul h
```

File: Mathlib/RingTheory/HahnSeries.lean [hunk 7]
```diff
@@ -512,6 +512,16 @@ instance : DistribMulAction R (HahnSeries Γ V) where
...
+  simp_all only [order, dite_false]
+  exact Set.IsWF.min_of_subset_not_lt_min (Function.support_smul_subset_right (fun _ => r) x.coeff)
+
+theorem smul_order_leq {Γ} [Zero Γ] [LinearOrder Γ] (r : R) (x : HahnSeries Γ V) (h : r • x ≠ 0) :
```
Comment (urkud) [inline Mathlib/RingTheory/HahnSeries.lean:522]:
```suggestion
theorem order_smul_le {Γ} [Zero Γ] [LinearOrder Γ] (r : R) (x : HahnSeries Γ V) (h : r • x ≠ 0) :
```

File: Mathlib/RingTheory/HahnSeries.lean [hunk 8]
```diff
@@ -512,6 +512,16 @@ instance : DistribMulAction R (HahnSeries Γ V) where
...
+    x.order ≤ (r • x).order := le_of_not_lt (smul_order_not_lt r x h)
```
Comment (urkud) [inline Mathlib/RingTheory/HahnSeries.lean:523]:
```suggestion
    x.order ≤ (r • x).order := le_of_not_lt (order_smul_not_lt r x h)
```

**Review by urkud — COMMENTED** [review]
Minor cleanup, otherwise LGTM.

File: Mathlib/RingTheory/HahnSeries.lean [hunk 9]
```diff
@@ -512,6 +512,15 @@ instance : DistribMulAction R (HahnSeries Γ V) where
     ext
     simp [mul_smul]
 
+theorem order_smul_not_lt [Zero Γ] (r : R) (x : HahnSeries Γ V) (h : r • x ≠ 0) :
+    ¬ (r • x).order < x.order := by
+  have hx : x ≠ 0 := right_ne_zero_of_smul h
+  simp_all only [order, dite_false]
+  exact Set.IsWF.min_of_subset_not_lt_min (Function.support_smul_subset_right (fun _ => r) x.coeff)
+
+theorem order_smul_le {Γ} [Zero Γ] [LinearOrder Γ] (r : R) (x : HahnSeries Γ V) (h : r • x ≠ 0) :
```
Comment (urkud) [inline Mathlib/RingTheory/HahnSeries.lean:521]: One last comment: `smul` is in the RHS, not LHS.
```suggestion
theorem le_order_smul {Γ} [Zero Γ] [LinearOrder Γ] (r : R) (x : HahnSeries Γ V) (h : r • x ≠ 0) :
```
Thanks!
bors d+

**Pull request closed without merging**